instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb

```diff
@@ -84,7 +84,7 @@ def span_attrs(kind, *args) # rubocop:disable Metrics/AbcSize
             end
 
             attrs = { 'db.operation' => validated_operation(operation), 'db.postgresql.prepared_statement_name' => statement_name }
-            attrs['db.statement'] = sql unless config[:db_statement] == :omit
+            attrs['db.statement'] = sql&.to_s unless config[:db_statement] == :omit
```

Comment:
Is there a circumstance where `sql` would be `nil`? That would have been problematic previously, since attribute values have to be non-nil, and validation would have dropped it and logged an error.

Reply:
Note: I realize this has been beaten to death, but I really don't think this level of defensiveness is warranted here.

Reply:
This test case 
https://github.com/open-telemetry/opentelemetry-ruby/blob/06b36a646afb30e86ad426f1be2bc97ed269fa15/instrumentation/pg/test/opentelemetry/instrumentation/pg/instrumentation_test.rb#L137-L138

WIthout the defensiveness it seems it's possible that lru_cache returns nil, which then gets encoded to `""` by `.to_s`. This failed test case is what prompted the defensiveness in the first place https://github.com/open-telemetry/opentelemetry-ruby/runs/5524602649?check_suite_focus=true

I hadn't really looked into the lru cache details or instrumentation implementation other than that it does seem possible, in some cases, for `sql` to be nil, we are currently testing for that case for some reason, and so we should have a `nil` check.

Reply:
It feels like that's more a case of the cache not working correctly. If the `lru_cache` returns `nil`, we should handle that cache miss and update the cache rather than papering over the problem with a `&.`.

Reply:
The `lru_cache` is a little unusual here - the expectation is that we write into the cache in the `PREPARE` case and read from it in the `EXECUTE` case. I think we should handle the `sql.nil?` case on line 81, where we read from the cache. By the time we get here, `sql` should be non-`nil`.

Reply:
I think the idea, (@ahayworth correct me if i'm wrong) is that this specific case is for handling prepared statements. Reviewing the test here, w/an lru cache of size `50`:

- user prepares 51 statements
- each time `prepare` is called, we perform the obfuscation and cache `<statement-name>:<obfuscated-prepared-statement>` k:v pair.
- The 51st call evicts the least recently used `<statement-name>:<obfuscated-prepared-statement>` k:v pair
- when the user later attempts to execute that already-evicted `<statement-name>:<obfuscated-prepared-statement>` k:v pair via `.execute(<statement-name>)`, it causes a cache miss _and_ we can't update the cache because we no longer have access to the `<obfuscated-prepared-statement>` to set as a value, so the lru_cache returns `nil`.

I don't know what the improvement would be here, this seems like an expected behavior, albeit one with tradeoffs.

Reply:
🤦 we handle `nil` attributes on the next line. This is complicated and confusing IMO. I think we should `.to_s` the result from `obfuscate_sql`, either in that method or on lines 72 and 77. Then we always cache strings.

Reply:
ok

Reply:
@ericmustin your understanding of the LRU cache is correct, both in functionality and in purpose.